stackrox/stackrox
# ROX-14948, ROX-17312: Improve rate-limited logger

## Description

The first implementation of the rate limited logger is sufficient for the case where central is running behind a traffic rate limiter.

In order to make it a more widely re-useable internal component that does not lose too much information about the traces submitted, the implementation is slightly altered in order to rate limit by limiter and trace, as well as count the number of suppressed traces and to issue the rate limited traces at some point.

## Checklist
~~- [ ] Evaluated and added CHANGELOG entry if required~~
~~- [ ] Determined and documented upgrade steps~~
~~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

If any of these don't apply, please comment below.

## Testing Performed

CI should be sufficient.

Some manual testing could be done by triggering frequent auth info extraction failures.

Manual testing:
- Create a goroutine that regularly issues rate-limited logs
```
go func rateLimitedLogTester() {
    ticker := time.NewTicker(1 * time.Second)
    for {
        select {
        case <-ticker.C:
            log.Info("This is a log that is issued every second")
            logging.GetRateLimitedLogger().InfoL("testlimiter", "This is a log that is issued every second to test the rate limited logger")
        }
    }
}()

```
- Check the logs issued by the rate-limited logger for the gorouting
```
main: 2023/06/30 16:35:22.316352 main.go:570: Info: This is a log that is issued every second
root logger: 2023/06/30 16:35:22.317236 logger.go:78: Info: central/main.go:571 - This is a log that is issued every second to test the rate limited logger
main: 2023/06/30 16:35:23.317389 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:24.316314 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:25.317176 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:26.315762 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:27.316280 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:28.315975 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:29.318751 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:30.317199 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:31.319104 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:32.316668 main.go:570: Info: This is a log that is issued every second
root logger: 2023/06/30 16:35:32.319106 logger.go:78: Info: central/main.go:571 - This is a log that is issued every second to test the rate limited logger - 9 log suppressed for limiter "testlimiter"
root logger: 2023/06/30 16:35:32.320221 logger.go:78: Info: central/main.go:571 - This is a log that is issued every second to test the rate limited logger
main: 2023/06/30 16:35:33.316331 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:34.318393 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:35.315804 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:36.316583 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:37.318287 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:38.315822 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:39.318019 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:40.316460 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:41.316767 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:42.317328 main.go:570: Info: This is a log that is issued every second
main: 2023/06/30 16:35:43.316800 main.go:570: Info: This is a log that is issued every second
root logger: 2023/06/30 16:35:43.317509 logger.go:78: Info: central/main.go:571 - This is a log that is issued every second to test the rate limited logger - 10 log suppressed for limiter "testlimiter"
root logger: 2023/06/30 16:35:43.317845 logger.go:78: Info: central/main.go:571 - This is a log that is issued every second to test the rate limited logger
```
